Add slide cropping

Philbrick opened the pull request:
- Add slide cropping to crop downsamples to create uniformity in pixel spacing across pyramid levels
- Adds end to end tests

Philbrick committed:
Add slide cropping

Add slide cropping to crop downsamples to create uniformity in pixel spacing across pyramid levels
Adds end to end tests

Philbrick committed:
fix style

Philbrick committed:
Add unit tests for wsiToDcm.

Refactor wsiToDcm to support unit testing.
Add Unit tests for:
  * getSlideLevelDim
  * getDownsampledLevelDimensionMM
  * getSmallestSlideDim
  * getOpenslideLevelForDownsample

danielbeaudreau commented on src/main.cpp:
```diff
@@ -140,7 +141,14 @@ int main(int argc, char *argv[]) {
+        ("uniformPixelSpacing",
+        programOptions::bool_switch(
+        &cropFrameToGenerateUniformPixelSpacing)->default_value(false),
+        "Crop imaging to generate downsampled mages with unifrom pixel "
```
typo

Philbrick replied:
fixed

danielbeaudreau commented on src/wsiToDcm.cpp:
```diff
@@ -172,6 +220,7 @@ std::unique_ptr<SlideLevelDim> WsiToDcm::getSlideLevelDim(const int32_t level,
       downsampleOfLevel = static_cast<double>(downsample) / multiplicator;
       // check that downsampling is going from higher to lower magnification
       if (downsampleOfLevel >= 1.0) {
+        levelToGet = priorLevel->level;
```
Why do we need to set levelToGet, levelWidth, or levelHeight here when they get set below? (line 233)

Just wondering if you've tested the scenario where we specify downsamples in the wrong order (decreasing).
Seems like we would just downsample based on the top level for each downsampled level.

Philbrick replied:
levelToGet and the other parameters are set here when the slide is being progressive downsampled.  generateUsingOpenSlide = false so the next if loop won't be entered.  Some of these are not actually used if progressive downsampling is being used. I set them here largely for completeness to a avoid undefined weirdness.

Yes, your comment, about level zero made me realize that the downsampling list enabled downsamples to be ordered in any order.  This would have broken the code.  The most recent CL should address this edge case and add unit tests to test it. Its not something that my usage of the code needs. But, should make it complete for others.

danielbeaudreau commented on src/wsiToDcm.cpp:
```diff
@@ -252,70 +312,102 @@ double  WsiToDcm::getDownsampledLevelDimensionMM(
+      firstSlideHeightCrop >= firstLevelHeight_) {
+      BOOST_LOG_TRIVIAL(info) << "Error Pixel spacing normalization " <<
+                         "cropped entire input image. Cannot generate DICOM.";
+      return 0;
```
should it be return 1 for an error

Philbrick replied:
fixed

danielbeaudreau commented on src/wsiToDcm.cpp:
```diff
@@ -252,70 +312,102 @@ double  WsiToDcm::getDownsampledLevelDimensionMM(
+                              firstSlideWidthCrop <<
+                              ", " <<
+                              firstSlideHeightCrop;
+  if (firstSlideWidthCrop >= firstLevelWidth_ ||
+      firstSlideHeightCrop >= firstLevelHeight_) {
+      BOOST_LOG_TRIVIAL(info) << "Error Pixel spacing normalization " <<
```
Is this condition possible?

Philbrick replied:
It shouldn't be.  removing.

danielbeaudreau commented on src/wsiToDcm.cpp:
```diff
@@ -252,70 +312,102 @@ double  WsiToDcm::getDownsampledLevelDimensionMM(
+                                          getSmallestSlideDim(&slideLevels));
+  if (wsiRequest_->cropFrameToGenerateUniformPixelSpacing) {
+    std::unique_ptr<SlideLevelDim> firstSlideLevel =
+                  std::move(getSlideLevelDim(0, NULL, smallestSlideDim.get()));
```
Is the slide at index 0 always guaranteed to be the largest?

Philbrick replied:
No, great catch.

Philbrick replied:
Fixed.

danielbeaudreau commented on src/wsiToDcm.cpp:
```diff
@@ -472,12 +568,14 @@ int WsiToDcm::dicomizeTiff() {
     higherMagnifcationDicomFiles.set_dicom_files(
                                            std::move(generatedDicomFiles));
+    // The combination of cropFrameToGenerateUniformPixelSpacing &
+    // can result in cropping images which span multiple frames/times
```
nit: comment seems incomplete, combination of cropping and _?

Philbrick replied:
fixed.

danielbeaudreau commented on tests/wsiToDcm.cpp:
```diff
@@ -136,3 +136,252 @@ TEST(compressionString, unknown) {
+  ASSERT_EQ(slide_dim->levelFrameHeight, 100);
+  ASSERT_EQ(slide_dim->frameWidthDownsampled, 200);
+  ASSERT_EQ(slide_dim->frameHeightDownsampled, 200);
+  ASSERT_EQ(slide_dim->frameWidthDownsampled, 200);
+  ASSERT_EQ(slide_dim->frameHeightDownsampled, 200);
```
nit: assertion is done twice

Philbrick replied:
removed duplicates

danielbeaudreau commented on tests/wsiToDcm.cpp:
```diff
@@ -136,3 +136,252 @@ TEST(compressionString, unknown) {
+  ASSERT_EQ(slide_dim->frameHeightDownsampled, 200);
+  ASSERT_EQ(slide_dim->frameWidthDownsampled, 200);
+  ASSERT_EQ(slide_dim->frameHeightDownsampled, 200);
+  ASSERT_EQ(slide_dim->cropSourceLevelWidth, 2220 - 2208);
+  ASSERT_EQ(slide_dim->cropSourceLevelHeight, 2967 - 2944);
+  ASSERT_TRUE(slide_dim->readOpenslide);
```
nit: could you add a comment explaining where 2208 and 2944 come from? e.g. 2208 = lowest dim (69) * largest downsample (32)

Philbrick replied:
Added.

Philbrick committed:
Update main.cpp

fix typo

Philbrick committed:
Update wsiToDcm.cpp

fix return error

Philbrick committed:
Update wsiToDcm.cpp

removed duplicate lines

Philbrick committed:
Update wsiToDcm.cpp

Add comment

Philbrick committed:
Update wsiToDcm.cpp

fix comment

Philbrick committed:
Fix bugs associated with the assumption that levels specified in order from highest to lowest magnification (least to most downsampled).

Old code made assumptions that first level was the largest image.  However Downsamples command line parameter enables caller to define downsampling in any order breaking this assumption.

- Changed variable names from firstLevelWidth_ to largestSlideLevelWidth_
- Changed  getSlideLevelDim first parameter level = -1 to force method to return dimensions for largestopenslideslide slice
-  Fixed getSmallestSlideDim to correctly identify the smallest slice when slices are defined out of order and stopDownsamplingAtSingleFrame = True
- Added error checking to test for cropping and user defined downsampling which are not a fixed multiples.
- Orders level processing to enable progressive downsampling for levels defined out of order.
- Adds error checking for the combination of cropping with no retileing.
- Adds unit tests testing out of order downsampling.

danielbeaudreau commented on src/wsiToDcm.cpp:
```diff
@@ -329,33 +371,95 @@ std::unique_ptr<SlideLevelDim>  WsiToDcm::getSmallestSlideDim(
+           (temp_downsample > smallsest_downsample)) ||
+          (smallest_level_is_single_frame &&
+           wsiRequest_->stopDownsamplingAtSingleFrame &&
+           frameCount == 1 && temp_downsample < smallsest_downsample)) {
```
nit: this condition is hard to read. could you pull out the compound conditions as bool values defined above and then use those in the if statement argument

Philbrick replied:
Done.

danielbeaudreau commented on src/wsiToDcm.cpp:
```diff
@@ -321,6 +359,10 @@ std::unique_ptr<SlideLevelDim>  WsiToDcm::getSmallestSlideDim(
   }
+  bool smallest_level_is_single_frame = false;
+  std::vector<std::tuple<int32_t, int64_t>> levelProcessOrder;
+  int64_t smallsest_downsample;
```
typo

Philbrick replied:
Done.

danielbeaudreau commented on src/wsiToDcm.cpp:
```diff
@@ -312,6 +336,20 @@ double  WsiToDcm::getDownsampledLevelDimensionMM(
+  const int64_t iDownsample = std::get<1>(i);
+  const int64_t jDownsample = std::get<1>(j);
+  if (iDownsample != jDownsample) {
+    // Sorta Downsampel : e/g/. Area 1, 2, 4, 8. 16
```
typo

danielbeaudreau commented on src/wsiToDcm.cpp:
```diff
@@ -329,33 +371,95 @@ std::unique_ptr<SlideLevelDim>  WsiToDcm::getSmallestSlideDim(
+        // if ((not stopDownsamplingAtSingleFrame or haven't seen single
+        // frame yet) and dimensions are smaller than previous frame) or
+        //
+        // if smallest area is single frame a stop downsampling at single
```
typo: "a" should be "and"?

Philbrick replied:
Done.

danielbeaudreau commented on src/wsiToDcm.cpp:
```diff
@@ -329,33 +371,95 @@ std::unique_ptr<SlideLevelDim>  WsiToDcm::getSmallestSlideDim(
-        if (frameX <= 1 && frameY <= 1) {
+        if (frameCount <= 1) {
+          smallest_level_is_single_frame = true;
+        }
```
nit: not sure if the encasing if statement is necessary here

Philbrick replied:
Moved up a level.  Done.

danielbeaudreau commented on src/wsiToDcm.cpp:
```diff
@@ -98,14 +97,34 @@ WsiToDcm::WsiToDcm(WsiRequest *wsiRequest) : wsiRequest_(wsiRequest) {
   }
   svsLevelCount_ = openslide_get_level_count(osr_);
   openslide_get_level_dimensions(osr_, 0,
```
should this be -1 too?

Philbrick replied:
No.  Its a Openslide API call.  Added comment.

danielbeaudreau commented on tests/wsiToDcm.cpp:
```diff
@@ -385,3 +450,67 @@ TEST(getSlideLevelDim, cropping_progressive) {
+  smallestSlideDim = std::move(converter.getSmallestSlideDim(&levels));
+  ASSERT_EQ(levels.size(), 6);
+  // -1 = get largest slice.
+  ASSERT_EQ(smallestSlideDim->levelWidthDownsampled, 69);
```
nit: should we add an assertion for level values similar to the other test case?

Philbrick replied:
Yes.  Done.

Philbrick committed:
Fix typo's and cleanup

danielbeaudreau commented:
/gcbrun

danielbeaudreau commented on src/wsiToDcm.cpp:
```diff
@@ -252,70 +337,183 @@ double  WsiToDcm::getDownsampledLevelDimensionMM(
+      levelProcessOrder.push_back(tpl);
+      if (wsiRequest_->stopDownsamplingAtSingleFrame && frameCount <= 1) {
+        BOOST_LOG_TRIVIAL(debug) << "Level[" <<level << "] frames:" <<
+                                    frameX << ", " << frameY;
```
nit: this debug log message isn't very useful as it will just log for the one frame with 1 x 1 dimension. Should we move it out of the if statement?

Philbrick replied:
done

Philbrick committed:
Update wsiToDcm.cpp

moved debug log

danielbeaudreau commented:
/gcbrun
